src/blog/2023/09/making-sense-of-your-industrial-data.md

```diff
@@ -0,0 +1,85 @@
+description: A look at legacy industrial protocol data and making sense of it in an IIoT world.
+date: 2023-09-08
+authors: ["stephen-mclaughlin"]
+image: "/blog/2023/09/images/industrial-legacy-data-blog-image.png"
+tags:
+    - posts
+    - node-red
+    - community
+---
+
+In today's rapidly evolving industrial landscape, the ability to collect, analyse, and act upon data is more critical than ever. The Industrial Internet of Things (IIoT) has ushered in a new era of connectivity and automation, but what about the wealth of data locked away in legacy industrial protocols like ModBus? How can we unlock and make sense of this valuable information in the IIoT era? That's where tools like Node-RED and node-red-contrib-buffer-parser come into play.
+
+<!--more-->
+
+### Legacy Industrial Data: ModBus
+
+ModBus is one of the most widely used industrial protocols. Originally developed in the late 1970s, it has been a popular choice for industrial communication ever since. However, its data format can be challenging to work with in the context of modern IIoT applications. ModBus typically represents data in 16-bit unsigned registers, making it necessary to convert this data into more usable formats like Signed integer, Float, Signed and Unsigned Long, String, or even individual bits.
+
+### A short primer on data types
+
+Before we dive into how to make sense of ModBus data, let's take a quick look at some of the common data types we have to deal with.
+
+#### 16-bit unsigned
+
+16-bit unsigned data is an integer that can only be positive. It can represent values from 0 to 65535. For example, the number 12345 is represented as `0x3039` in hexadecimal or `0011000000111001` in binary.
+
+#### 16-bit signed
+
+16-bit signed data is an integer that can be positive or negative. It can represent values from -32768 to 32767. For example, the number -12345 is represented as `0xCFC7` in hexadecimal or `1100111111000111` in binary.
+
+#### 32-bit data
+
+32-bit data, like 16-bit data can mean many things. It could be signed or unsigned, or even a floating point number. For example, the number 12345 is represented as `0x00003039` in hexadecimal or `00000000000000000011000000111001` in binary. Typically, 32-bit data is represented as two 16-bit registers. Therefore, when dealing with 32-bit data, you need to combine two 16-bit registers to get the full value.
+
+#### Endianness
+
+Endianness is a term used to describe the order in which bytes are stored in memory. There are two types of endianness: big-endian and little-endian. In big-endian, the most significant byte is stored first, while in little-endian, the least significant byte is stored first. For example, the number 12345 is represented as `0x3039` in big-endian and `0x3930` in little-endian.  In other words, the bytes might be stored in reverse order. This further complicates the process of converting ModBus data into more usable formats.
```

Comment:
```suggestion
Endianness is a term used to describe the order in which bytes are stored in memory. There are two types of endianness: big-endian and little-endian. In big-endian, the most significant bit is stored first, while in little-endian, the least significant bit is stored first. For example, the number 12345 is represented as `0x3039` in big-endian and `0x3930` in little-endian.  In other words, the bits might be stored in reverse order. This further complicates the process of converting ModBus data into more usable formats.
```

The endianness is on a bit level, not byte level. :)

Reply:
I am afraid 100% acurate - in these terms of protocols.

The bytes are actually swapped not the bit ordering.

From the PoV of a buffer (of bytes), the functions doeing the work behind the scenes is the NodeJS buffer: https://nodejs.org/api/buffer.html#bufreadint16beoffset - the BE denoted BigEndian (byte swapped), the LE is ...

Other readme...

Stack overflow
> In modern computing endianness only applies to ordering of bytes, not bits

Wikipedia 
> In [computing](https://en.wikipedia.org/wiki/Computing), endianness is the order or sequence of [bytes](https://en.wikipedia.org/wiki/Byte) of a [word](https://en.wikipedia.org/wiki/Word_(data_type)) of digital data in [computer memory](https://en.wikipedia.org/wiki/Computer_memory) or [data communication](https://en.wikipedia.org/wiki/Data_communication) which is identified by describing the impact of the "first" bytes, meaning at the smallest address or sent first. Endianness is primarily expressed as big-endian (BE) or little-endian (LE). A big-endian system stores the [most significant byte](https://en.wikipedia.org/wiki/Most_significant_byte) of a word at the smallest [memory address](https://en.wikipedia.org/wiki/Memory_address) and the [least significant byte](https://en.wikipedia.org/wiki/Least_significant_byte) at the largest.


Some other site
>BE stores the big-end first. When reading multiple bytes the first byte (or the lowest memory address) is the biggest - so it makes the most sense to people who read left to right.
>LE stores the little-end first. When reading multiple bytes the first byte (or the lowest memory address) is the littlest -  so it makes most sense to people who read right to left.

Reply:
Thanks, I didn't google, but would've bet my house on it just a few minutes ago. Today I learned (again)